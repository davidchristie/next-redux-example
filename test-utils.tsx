import React from "react";
import { Provider } from "react-redux";
import { createStore, Dispatch, Store, Action } from "redux";
import { render } from "@testing-library/react";
import { exampleInitialState, RootState } from "./reducer";

type Context = ReturnType<typeof renderConnectedComponent>;

type GetContext = () => Context;

const createWrapper = (initialState: RootState) => {
  const reducer = (state = initialState) => state;
  const store = createStore(reducer);
  const dispatch = mockDispatch(store);
  const Wrapper: React.FunctionComponent = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  );
  return { dispatch, Wrapper };
};

const mockDispatch = (store: Store): jest.Mock<Dispatch> => {
  const dispatch = jest.fn(store.dispatch);
  store.dispatch = dispatch;
  return dispatch;
};

const renderConnectedComponent = (
  element: React.ReactElement,
  { initialState = exampleInitialState } = {}
) => {
  const { dispatch, Wrapper } = createWrapper(initialState);
  const result = render(element, {
    wrapper: Wrapper
  });
  return { dispatch, result };
};

export const describeComponent = <P extends {}>(
  Component: React.ComponentType<P>,
  props: P,
  fn: (getContext: GetContext) => void
) => {
  describe(`${Component.displayName} Component`, () => {
    let context: Context;

    beforeEach(() => {
      context = renderConnectedComponent(<Component {...props} />);
    });

    it("Has display name", () => {
      expect(Component.displayName).toBeDefined();
    });

    fn(() => context);
  });
};

export const describeComponentWithReduxState = <P extends {}>(
  Component: React.ComponentType<P>,
  props: P,
  state: Partial<RootState>,
  fn: (getContext: GetContext) => void
) => {
  describe(`${Component.displayName} with Redux state: ${JSON.stringify(
    state
  )}`, () => {
    let context: Context;

    beforeEach(() => {
      context = renderConnectedComponent(<Component {...props} />, {
        initialState: { ...exampleInitialState, ...state }
      });
    });

    fn(() => context);
  });
};

export const describeWhenButtonIsClicked = (
  getContext: GetContext,
  text: string,
  fn: (getContext: GetContext) => void
) => {
  describe(`When "${text}" button is clicked`, () => {
    let context: Context;

    beforeEach(() => {
      context = getContext();
      const button = context.result.getByText(text, {
        selector: "button"
      });
      button.click();
    });

    fn(() => context);
  });
};

export const itHasLink = (
  getContext: GetContext,
  text: string,
  href: string
) => {
  it(`Has link "${text}" to "${href}"`, () => {
    const { result } = getContext();
    expect(
      result.getByText(text, {
        selector: `a[href="${href}"]`
      })
    ).not.toBeNull();
  });
};

export const itShouldDispatchAction = (
  getContext: GetContext,
  action: Action
) => {
  it(`Dispatches "${action.type}" action`, () => {
    const { dispatch } = getContext();
    expect(dispatch).toBeCalledWith(action);
  });
};

export const itShouldMatchSnapshot = (getContext: GetContext) => {
  it("Should match snapshot", () => {
    const { result } = getContext();
    expect(result.asFragment()).toMatchSnapshot();
  });
};

export const itShowsText = (getContext: GetContext, text: string) => {
  it(`Shows "${text}"`, () => {
    const { result } = getContext();
    expect(result.getByText(text)).not.toBeNull();
  });
};
