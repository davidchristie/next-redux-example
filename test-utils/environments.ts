const environments = ["development", "production", "test"];

export const describeAllEnvironments = (fn: (environment: string) => void) => {
  const originalEnv = process.env;

  beforeEach(() => {
    jest.resetModules();
    const mockEnv = { ...originalEnv };
    delete mockEnv.NODE_ENV;
    process.env = mockEnv;
  });

  afterEach(() => {
    process.env = originalEnv;
  });

  describe.each(environments.map(environment => [environment]))(
    'In Environment: "%s"',
    environment => {
      (process.env as any).NODE_ENV = environment;

      fn(environment);
    }
  );
};
