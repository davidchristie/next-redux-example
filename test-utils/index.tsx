export * from "./components";
export * from "./environments";
export * from "./objects";
export * from "./reducers";
export * from "./stores";
