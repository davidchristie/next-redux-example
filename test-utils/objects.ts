export const itShouldHaveMethod = <T>(instance: T, method: keyof T) => {
  it(`Should have ${method} method`, () => {
    expect(instance[method]).toBeInstanceOf(Function);
  });
};
