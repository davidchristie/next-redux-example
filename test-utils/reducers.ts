import { Reducer, AnyAction } from "redux";

export const itShouldHaveNextState = <S extends {}>(
  name: string,
  {
    action,
    expectedNextState,
    previousState,
    reducer
  }: {
    action: AnyAction;
    expectedNextState: S;
    previousState: S;
    reducer: Reducer<S>;
  }
) => {
  it(name, () => {
    const actualNextState = reducer(previousState, action);
    expect(actualNextState).toEqual(expectedNextState);
  });
};
