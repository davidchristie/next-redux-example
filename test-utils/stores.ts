import { Dispatch, Store } from "redux";

export const mockDispatch = (store: Store): jest.Mock<Dispatch> => {
  const dispatch = jest.fn(store.dispatch);
  store.dispatch = dispatch;
  return dispatch;
};
