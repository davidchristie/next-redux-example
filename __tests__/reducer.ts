import {
  loadDataSuccess,
  decrement,
  failure,
  increment,
  reset,
  tickClock
} from "../actions";
import rootReducer, { exampleInitialState } from "../reducer";
import { itShouldHaveNextState } from "../test-utils";

describe("Root Reducer", () => {
  describe("Initial state", () => {
    it("Should match snapshot", () => {
      const initialState = rootReducer(undefined, { type: "@@INIT" });
      expect(initialState).toMatchSnapshot();
    });
  });

  describe("After Failure", () => {
    const error = new Error("TEST ERROR");

    itShouldHaveNextState("Should store the error", {
      action: failure(error),
      expectedNextState: {
        ...exampleInitialState,
        error
      },
      previousState: exampleInitialState,
      reducer: rootReducer
    });
  });

  describe("After Increment", () => {
    itShouldHaveNextState("Should increase count by 1", {
      action: increment(),
      expectedNextState: {
        ...exampleInitialState,
        count: exampleInitialState.count + 1
      },
      previousState: exampleInitialState,
      reducer: rootReducer
    });
  });

  describe("After Decrement", () => {
    itShouldHaveNextState("Should reduce count by 1", {
      action: decrement(),
      expectedNextState: {
        ...exampleInitialState,
        count: exampleInitialState.count - 1
      },
      previousState: exampleInitialState,
      reducer: rootReducer
    });
  });

  describe("After Reset", () => {
    itShouldHaveNextState("Should set count to initial value", {
      action: reset(),
      expectedNextState: {
        ...exampleInitialState,
        count: exampleInitialState.count
      },
      previousState: {
        ...exampleInitialState,
        count: 123
      },
      reducer: rootReducer
    });
  });

  describe("After Data Load Success", () => {
    const placeholderData = [{ id: 1 }, { id: 2 }, { id: 3 }];

    itShouldHaveNextState("Should store the data", {
      action: loadDataSuccess(placeholderData),
      expectedNextState: {
        ...exampleInitialState,
        placeholderData
      },
      previousState: exampleInitialState,
      reducer: rootReducer
    });
  });

  describe("After Clock Tick", () => {
    const timestamp = 1580554431955;

    describe("On Client", () => {
      const isServer = false;

      itShouldHaveNextState("Should update the timestamp and turn on light", {
        action: tickClock(timestamp, isServer),
        expectedNextState: {
          ...exampleInitialState,
          lastUpdate: timestamp,
          light: true
        },
        previousState: exampleInitialState,
        reducer: rootReducer
      });
    });

    describe("On Server", () => {
      const isServer = true;

      itShouldHaveNextState("Should update the timestamp and turn off light", {
        action: tickClock(timestamp, isServer),
        expectedNextState: {
          ...exampleInitialState,
          lastUpdate: timestamp,
          light: false
        },
        previousState: exampleInitialState,
        reducer: rootReducer
      });
    });
  });
});
