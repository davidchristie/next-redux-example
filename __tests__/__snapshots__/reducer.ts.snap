// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root Reducer Initial state Should match snapshot 1`] = `
Object {
  "count": 0,
  "error": undefined,
  "lastUpdate": 0,
  "light": false,
  "placeholderData": null,
}
`;
